### PR TITLE
EKS support

### PR DESCRIPTION
**Pull request by airman604**
Added EKS support based on the EKS bootstrap script - https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh

For EKS, it is possible to authenticate as the node on which the pod is running using kubeconfig template from the repository above (https://github.com/awslabs/amazon-eks-ami/blob/master/files/kubelet-kubeconfig) and information from user-data metadata endpoint.

EKS mode is added as a top-level command as it doesn't require a two-step (bootstrap/generate) invocation. It will generate kubeconfig file and automatically download needed aws-iam-authenticator.

As a side note - I couldn't get kubeletmein (even without EKS support) compiled in golang 1.16, I think it's due to changes in module handling, but don't have enough expertise/patience to figure out how to fix it. Builds fine under golang 1.15

### CODE REVIEW

**Comment by marcwickenden** [comment]
Firstly, thank you so much for taking the time to work on this and submit a PR. I really appreciate that.

I need to have a proper read through of the code before I merge it because you've taken a different approach to the one I originally intended however, I can see that this might be the most effective route.

One thing I have spotted is the downloading of `aws-iam-authenticator`. I don't want `kubeletmein` to be responsible for downloading other external tools. I understand why you have done this, as a convenience, however I subscribe to the Unix philosophy of "do one thing and do it well". If we start to incorporate external tool requirements we'd have to maintain that and worry about versions/compatibility, etc.

Finally, I've just quickly add go 1.16 support and removed the `vendor` directory. If you merge off master or the latest release you should hopefully find that now works. The v0.7.2 binary under Releases was just built using 1.16 so it seems ok.

**Comment by airman604** [comment]
Hi Marc, I can remove downloading of `aws-iam-authenticator` and add instructions to download it to the output instead - that makes sense. Will also check everything works under go 1.16

**Comment by airman604** [comment]
Done. Also, happy to investigate other options for this functionality if needed.

**Comment by marcwickenden** [comment]
Hoping to merge this in tomorrow. Am just doing some refactoring and then will pull it in I think. Thank you again.

**Comment by airman604** [comment]
No problem, and thank **you** for doing the research and writing the tool in the first place!

**Comment by marcwickenden** [comment]
Which version of EKS did you run this on? I don't see the nodes being bootstrapped in this way with 1.18. user-data is gzipped and provides the kubelet config file in there without any other processing required. No need for `aws-iam-authenticator` either, it uses the aws cli. Just need to test for both.

**Comment by airman604** [comment]
Tested on 1.18 using Amazon Linux for the nodes. `user-data` I am getting from within a pod is a MIME enveloped script that sets up environment variables and calls `bootstrap.sh`, here's an example:
```
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="//"

--//
Content-Type: text/x-shellscript; charset="us-ascii"
#!/bin/bash
set -ex
B64_CLUSTER_CA=LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUN5RENDQWJDZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREFWTVJNd0VRWURWUVFERXdwcmRXSmwKY201bGRHVnpNQjRYRFRJeE1ETXdNakl4TWpBMU4xb1hEVE14TURJeU9ESXhNakExTjFvd0ZURVRNQkVHQTFVRQpBeE1LYTNWaVpYSnVaWFJsY3pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTEZmClBuWmNkNmVCd29MYnpyNDR6OFpZU3ZJUzd5UGd4WGN6NkxveCtKc0llQnptTUppckdkaGhRaFVaclZjYjZpZ0EKNHk0djBsaEQxeU1Rb042UzJBd0haWlpjeXBXU0xLWnczNE93aGo1bzA3cFR1ektUU0R3MHpBblk1Ni9YWERYLwpaOUFSVGlWRjBDTUYydG9Md0FMNlQ5SG9wRzhMVlJXeEp6Z3daemtZK1FtbDhEM1FKZzNsS3NGNU14cy9kNG5MCmVXdGsrRDJmRmkvWTVDN054bk1XaEJwMWo5OG9xYWVKTTdtclkxMkIvRnQwUGk5RHAxSVVaY28zNTJ4U0tHSHMKajdrMEN1cG0rM0taY2ZLVnZ1aTQ5M2phdlRVdFJEbEFRbGZyMUIraGttNDBweDh4V1h2ZWt4aW1Ya1grUjl0YQpVemo0elZ5cThoUlhHUEtNWk9rQ0F3RUFBYU1qTUNFd0RnWURWUjBQQVFIL0JBUURBZ0trTUE4R0ExVWRFd0VCCi93UUZNQU1CQWY4d0RRWUpLb1pJaHZjTkFRRUxCUUFEZ2dFQkFGNTVaS2owMXRySmVqeDBKY0hOeGdTckR3R2EKbDQrUTBERW51NXNOQUNrRzJuSmRqOGNnUVR4cFBENkN4M1VRMnNESkVlZEtMSUtiNkZRNnZUZWtvbGI3aGloRAptV01aOFlyY2VhT0ZsZkJ6QWNHcEhxMk41c0xxL3JMdzFtVkpHR0lNc2x5d2VBbC9YTUsycms5aFJIZVh2dk1RCmlDL2wrUjFGWXYrZG9GYmwzZ0R6em40REtCMDNMR3ZQSHBWdnprUzlYOGp0M2gyWGNkeW9aVjAyRzJ0MC91M0IKeWFNbGlIb1FkNGFuZnQybUwyUEFaa015TVIwcFN5UU9OWFZEY3FRQzF5cG1JNDRFS0xxU0ljam1hWWg0dUkxNgprM0sweDJNVVAwS3lyVkRyWXFLWWJxWXNqaC9hTE0zdUhjaXV1SDRGL21IdDJHM2ZTcTJTUk9qRTdOMD0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=
API_SERVER_URL=https://E59E8F4FA25161F2A37857B3B5D3582D.gr7.us-west-2.eks.amazonaws.com
K8S_CLUSTER_DNS_IP=10.100.0.10
/etc/eks/bootstrap.sh shiny-new-cluster --kubelet-extra-args '--node-labels=eks.amazonaws.com/nodegroup-image=ami-09729f03d2b24317c,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/nodegroup=node-group-1' --b64-cluster-ca $B64_CLUSTER_CA --apiserver-endpoint $API_SERVER_URL --dns-cluster-ip $K8S_CLUSTER_DNS_IP

--//--
```

Yes, the `kubeconfig` uses instance identity for authentication, kubectl gets token by running an external command, which can be either `aws-iam-authenticator`, or AWS CLI. As `aws-iam-authenticator` is a simpler download (single go binary), I thought it made sense to use that. For more details see **Create kubeconfig manually** here: https://docs.aws.amazon.com/eks/latest/userguide/create-kubeconfig.html
Bottom line is, `aws-iam-authenticator token -i <cluster name>` and `aws eks get-token --cluster-name <cluster name>` do the same thing. The information from `user-data` that is needed for `kubeconfig` are kube-api endpoint, CA certificate and the cluster name.

**Comment by marcwickenden** [comment]
Ok, I see that now on a second cluster. It has to have managed nodes to have that user-data value. If the nodes aren't managed then it uses `cloud-config` format. I will update to handle this.

**Comment by airman604** [comment]
Are you provisioning your cluster with eksctl?

It seems like `user-data` is different depending on the provisioning method - there's also a possibility that the cluster was provisioned through UI, but (unmanaged) nodes are provisioned through CloudFormation (https://docs.aws.amazon.com/eks/latest/userguide/launch-workers.html).

With the exception of the cluster name, all the information for the `kubeconfig` can be retrieved with aws cli (see line 305 in `bootstrap.sh` here: https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh). That may be the most flexible way to go - require aws cli as a dependency, then try to extract cluster name from user-data using different methods (gzip uncompress and look for calls to bootstrap.sh), and having the cluster name get all the other information using aws cli.

**Comment by marcwickenden** [comment]
Yep, already got it implemented in test. Just fighting with k8s go dependencies and then I should have a working release to cover both scenarios.
